test/unittest/datasets/data_producer_common_tests.cpp

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file data_producer_common_tests.cpp
+ * @date 12 July 2021
+ * @brief Common test for nntrainer data producers (Param Tests)
+ * @see	https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+#include <gtest/gtest.h>
+
+#include <data_producer_common_tests.h>
+
+void DataProducerSemantics::SetUp() {
+  auto [producerFactory, properties, input_dims_, label_dims_, validator_,
+        result_] = GetParam();
+
+  producer = producerFactory(properties);
+  input_dims = std::move(input_dims_);
+  label_dims = std::move(label_dims_);
+  result = result_;
+  validator = std::move(validator_);
+
+  if (result != DataProducerSemanticsExpectedResult::SUCCESS) {
+    ASSERT_EQ(validator, nullptr)
+      << "Given expected result of not success, validator must be empty!";
+  }
+}
+
+void DataProducerSemantics::TearDown() {}
+
+TEST_P(DataProducerSemantics, finalize_pn) {
+  if (result == DataProducerSemanticsExpectedResult::FAIL_AT_FINALIZE) {
+    EXPECT_ANY_THROW(producer->finalize(input_dims, label_dims));
+  } else {
+    EXPECT_NO_THROW(producer->finalize(input_dims, label_dims));
+  }
+}
+
+TEST_P(DataProducerSemantics, fetch_one_epoch_or_10_iteration_pn) {
+  if (result != DataProducerSemanticsExpectedResult::SUCCESS) {
+    return; // skip this test
+  }
+
+  auto generator = producer->finalize(input_dims, label_dims);
+  auto sz = producer->size(input_dims, label_dims);
+  bool has_fixed_size = sz != nntrainer::DataProducer::SIZE_UNDEFINED;
+
+  if (!has_fixed_size) {
+    sz = 5;
+  }
+
+  for (unsigned i = 0; i < sz; ++i) {
+    auto [last, ins, labels] = generator();
+
+    ASSERT_FALSE(last) << " reached last at iteration: " << i << '\n';
```

Comment:
why use ASSERT than EXPECT?

Reply:
If last is true, there is no meaning proceeding, while producing too many error logs making it hard to identify, where it was actually dead.

Reply:
Then add a manual if. In my understanding, ASSERT is used for setting up the test case, but not the actual test itself.
In this case, it seems like you put this to avoid printing all the mismatch in the loops and just print the first, but IMO, the printing all can be helpful, as that information is helpful (it is possible for a random piece of data to mismatch but others do match and these observations hints what the error can be in general).

Reply:
Continuing doesn't make sense after this line fails. Same things applied already to unittest_models. 
Adding a manual will do exactly same. 

Please refer to below.

http://google.github.io/googletest/primer.html#assertions

> Usually EXPECT_* are preferred, as they allow more than one failure to be reported in a test. However, you should use ASSERT_* if it doesn’t make sense to continue when the assertion in question fails.

Reply:
I guess its our choice, either works :).